### PR TITLE
fix(har): restart redirected navigation

### PR DESCRIPTION
**Pull request by yury-s**

### CODE REVIEW

File: packages/playwright-core/src/server/frames.ts [hunk 1]
```diff
@@ -603,48 +609,71 @@ export class Frame extends SdkObject {
...
-      const headers = this._page._state.extraHTTPHeaders || [];
-      const refererHeader = headers.find(h => h.name.toLowerCase() === 'referer');
-      let referer = refererHeader ? refererHeader.value : undefined;
-      if (options.referer !== undefined) {
-        if (referer !== undefined && referer !== options.referer)
-          throw new Error('"referer" is already specified as extra HTTP header');
-        referer = options.referer;
-      }
-      url = helper.completeUserURL(url);
+      while (true) {
```
Comment (pavelfeldman) [inline packages/playwright-core/src/server/frames.ts:612]: I don't feel good about while true in _goto (our core functionality) over a redirect in har (our p3 functionality). Can we do client-side redirect instead?

File: packages/playwright-core/src/client/network.ts [hunk 2]
```diff
@@ -282,9 +282,9 @@ export class Route extends ChannelOwner<channels.RouteChannel> implements api.Ro
     this._reportHandled(false);
   }
 
-  async abort(errorCode?: string) {
```
Comment (yury-s) [inline packages/playwright-core/src/client/network.ts:285]: make this param private

File: packages/playwright-core/src/server/dispatchers/frameDispatcher.ts [hunk 3]
```diff
@@ -65,6 +65,10 @@ export class FrameDispatcher extends Dispatcher<Frame, channels.FrameChannel> im
     });
   }
 
+  async redirectNavigationAfterAbort(params: channels.FrameRedirectNavigationAfterAbortParams, metadata: CallMetadata): Promise<channels.FrameRedirectNavigationAfterAbortResult> {
```
Comment (yury-s) [inline packages/playwright-core/src/server/dispatchers/frameDispatcher.ts:68]: revert

File: packages/playwright-core/src/server/frames.ts [hunk 4]
```diff
@@ -586,73 +601,94 @@ export class Frame extends SdkObject {
...
       this._page._disconnectedPromise.then(() => { throw new Error('Navigation failed because page was closed!'); }),
       this._page._crashedPromise.then(() => { throw new Error('Navigation failed because page crashed!'); }),
       this._detachedPromise.then(() => { throw new Error('Navigating frame was detached!'); }),
-      action(),
+      (async (): Promise<network.Response | null> => {
+        try {
+          return await action();
+        } catch (e) {
+          if (this._navigationRedirectedURL && e instanceof NavigationAbortedError && e.documentId === this._redirectedNavigationDocumentId) {
+            const url = this._navigationRedirectedURL;
```
Comment (yury-s) [inline packages/playwright-core/src/server/frames.ts:614]: always cleanup

**Review by pavelfeldman — COMMENTED** [review]
That still sounds way too complex, let's not land as is.

File: packages/playwright-core/src/server/frames.ts [hunk 5]
```diff
@@ -586,73 +598,90 @@ export class Frame extends SdkObject {
...
+            return this._gotoAction(progress, url, options);
+          }
+        }
+        throw e;
+      }),
     ]);
   }
 
+  redirectNavigationAfterAbort(url: string, documentId: string) {
+    assert(!this._pendingNavigationRedirectAfterAbort, `Already redirecting to ${this._pendingNavigationRedirectAfterAbort?.url}`);
```
Comment (dgozman) [inline packages/playwright-core/src/server/frames.ts:621]: Drop the assert :)

Comment (yury-s) [reply packages/playwright-core/src/server/frames.ts]: done.

**Review by pavelfeldman — APPROVED** [review]

**Pull request merged**